src/nimblepkg/nimscriptwrapper.nim

```diff
@@ -0,0 +1,131 @@
+# BSD License. Look at license.txt for more info.
+
+## Implements the new configuration system for Nimble. Uses Nim as a
+## scripting language.
+
+import common, version, options, packageinfo, cli
+import hashes, json, os, streams, strutils, strtabs, tables, times, osproc, sets, pegs
+
+type
+  Flags = TableRef[string, seq[string]]
+  ExecutionResult*[T] = object
+    success*: bool
+    command*: string
+    arguments*: seq[string]
+    flags*: Flags
+    retVal*: T
+
+const
+  internalCmd = "e"
+  nimscriptApi = staticRead("nimscriptapi.nim")
+  nimscriptHash = $nimscriptApi.hash().abs()
+
+proc execNimscript(nimsFile, actionName: string, options: Options, live = true): tuple[output: string, exitCode: int] =
+  var
+    cmd = ("nim e --hints:off --verbosity:0 " & nimsFile.quoteShell & " " & actionName).strip()
+
+  if live:
+    result.exitCode = execCmd(cmd)
+    let
+      outFile = nimsFile & ".out"
+    if outFile.fileExists():
+      result.output = outFile.readFile()
+      discard outFile.tryRemoveFile()
+  else:
+    result = execCmdEx(cmd, options = {poUsePath})
+
+proc setupNimscript*(scriptName: string, options: Options): tuple[nimsFile, iniFile: string] =
+  let
+    cacheDir = getTempDir() / "nimblecache"
+    shash = $(scriptName & nimscriptHash).hash().abs()
+    prjCacheDir = cacheDir / scriptName.splitFile().name & "_" & shash
+    nimsCacheFile = prjCacheDir / scriptName.extractFilename().changeFileExt ".nims"
+
+  result.nimsFile = scriptName.parentDir() / scriptName.splitFile().name & "_" & shash & ".nims"
+  result.iniFile = prjCacheDir / scriptName.extractFilename().changeFileExt ".ini"
+
+  if not prjCacheDir.dirExists() or not nimsCacheFile.fileExists() or not result.iniFile.fileExists() or
+    scriptName.getLastModificationTime() > nimsCacheFile.getLastModificationTime():
+    createDir(prjCacheDir)
+    writeFile(nimsCacheFile, nimscriptApi & scriptName.readFile() & "\nonExit()\n")
+    discard tryRemoveFile(result.iniFile)
+
+  if not result.nimsFile.fileExists():
+    nimsCacheFile.copyFile(result.nimsFile)
+
+  if not result.iniFile.fileExists():
+    let
+      (output, exitCode) = result.nimsFile.execNimscript("printPkgInfo", options, live=false)
+
+    if exitCode == 0 and output.len != 0:
+      result.iniFile.writeFile(output)
+    else:
+      raise newException(NimbleError, output & "\nprintPkgInfo() failed")
+
+proc execScript*(scriptName, actionName: string, options: Options): ExecutionResult[bool] =
+  let
+    (nimsFile, iniFile) = setupNimscript(scriptName, options)
+
+    (output, exitCode) = nimsFile.execNimscript(actionName, options)
+
+  defer:
+    nimsFile.removeFile()
+
+  if exitCode != 0:
+    raise newException(NimbleError, output)
+
+  let
+    j =
+      if output.len != 0:
+        parseJson(output)
+      else:
+        parseJson("{}")
+
+  result.flags = newTable[string, seq[string]]()
+  if "success" in j:
+    result.success = j["success"].getBool()
```

Comment:
These can be simplified to:

```nim
result.success = j{"success"}.getBool()
```

and the same for others.

Reply:
You could actually probably just use the `to` macro here.

Reply:
I'll defer this for now, don't want to spend more time on this, especially since there's a table in this object (flags).